sw/airborne/modules/flight_plan_guided/flight_plan_guided.c

```diff
@@ -0,0 +1,427 @@
+ * @param[in] in_flight, bool necessary for stabilization loop to check if the MAV is flying
+ * @param[return] Boolean for flight plan to keep running (true) or to just run once (false)
+ */
+bool ResetAngles_ATT(float current_heading, bool in_flight)
+{
+  struct Int32Eulers cmd;
+  if (guidance_h.mode == GUIDANCE_H_MODE_ATTITUDE) {
+    cmd.phi = ANGLE_BFP_OF_REAL(0.0f);
+    cmd.theta = ANGLE_BFP_OF_REAL(0.0f);
+    cmd.psi = ANGLE_BFP_OF_REAL(current_heading);
+
+    stabilization_attitude_set_rpy_setpoint_i(&cmd);
+    stabilization_attitude_run(in_flight);
+  }
+  return false;
+}
+
+/********************************************//**
+ *  Blocking conditions
+ ***********************************************/
+
+/**
+ * WaitUntilAltitude: Blocks while waiting until a certain altitude is reached
+ * @param[in] altitude in [m] (positive z-axis is up)
+ * @param[return] Boolean for flight plan to keep running (true) or to just run once (false)
+ */
+bool WaitUntilAltitude(float altitude)
+{
+  if (autopilot.mode != AP_MODE_GUIDED) { return true; }
+
+  if (stateGetPositionEnu_f()->z < -altitude) { return true; }
+
+  return false;
+}
+
+/**
+ * WaitUntilSpeedOrAltitude: Blocks while waiting until a certain altitude or speed is reached
+ * @param[in] fail_altitude, wait for altitude in [m] (positive z-axis is up)
+ * @param[in] speed, speed in [m/s] (positive z-axis is up)
+ * @param[return] Boolean for flight plan to keep running (true) or to just run once (false)
+ */
+bool WaitUntilSpeedOrAltitude(float speed, float fail_altitude)
+{
+  if (autopilot.mode != AP_MODE_GUIDED) { return true; }
+
+  if (stateGetPositionEnu_f()->z > -fail_altitude) { return false; }
+  if (stateGetSpeedEnu_f()->z < -speed) { return true; }
+
+  return false;
+}
+
+/**
+ * WaitforHeading: Blocks untill wanted heading is achieved
+ * @param[in] heading, waits untill MAV reached certain heading [rad] (clockwise is positive)
+ *  @param[return] Boolean for flight plan to keep running (true) or to just run once (false)
+ *
+ * TODO: Wrap the angle around -pi and pi
+ */
+bool WaitforHeading(float heading)
+{
+  if (fabs(heading - stateGetNedToBodyEulers_f()->psi) < 0.1) {
+    return false;
+  }
+  return true;
+}
+
+/**
+ * WaitForHMode: Blocks until specified horizontal mode is achieved
+ * @param[in] mode, number of the specific horizontal mode that you want to wait for
+ * @param[return] Boolean for flight plan to keep running (true) or to just run once (false)
+ */
+bool WaitForHMode(uint8_t mode)
+{
+  if (guidance_h.mode == mode) {
+    return false;
+  }
+  return true;
+}
+
+
+/********************************************//**
+ *  Timer functions
+ ***********************************************/
+
+/**
+ * ResetSpecialTimer: Reset a special timer, that works outside of the block and stage time
+ * @param[return] Boolean for flight plan to keep running (true) or to just run once (false)
+ */
+float specialtimer = 0;
```

Comment:
specialtimer is a bit special name, maybe nav_timer? 


As it is currently implemented, this timer does not run continuously but only when waitUintilTimer or  WaitUntilTimerOrAltitude is called. If you areally want a timer that is always running you should add it to common_flight_plan.h and navigation.c but only reset it with your ResetNavTimer.